### PR TITLE
Snapshot code refactoring

### PR DESCRIPTION
**Pull request by 766C6164**
This PR moves snapshot scheduler and snapshot creation logic/code outside of producer plugin and into the chain library. Chain library already contains snapshot writer so it should be a more natural fit for this code

Snapshot scheduling test is still present in producer plugin and will be probably moved away in a scope of another issue: https://github.com/AntelopeIO/leap/issues/586

### CODE REVIEW

File: libraries/chain/include/eosio/chain/snapshot_scheduler.hpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+   uint32_t _snapshot_id = 0;
+   uint32_t _inflight_sid = 0;
+
+   // path to write the snapshots to
+   bfs::path _snapshots_dir;
+
+   void x_serialize() {
+      auto& vec = _snapshot_requests.get<as_vector>();
+      std::vector<snapshot_scheduler::snapshot_schedule_information> sr(vec.begin(), vec.end());
```
Comment (greg7mdp) [inline libraries/chain/include/eosio/chain/snapshot_scheduler.hpp:98]: remove all the unnecessary `snapshot_scheduler::` within the class

File: libraries/chain/snapshot_scheduler.cpp [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+   // returning snapshot_schedule_result
+   return snapshot_scheduler::snapshot_schedule_result{{_snapshot_id - 1}, {sri.block_spacing, sri.start_block_num, sri.end_block_num, sri.snapshot_description}};
+}
+
+snapshot_scheduler::snapshot_schedule_result snapshot_scheduler::unschedule_snapshot(uint32_t sri) {
+   auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+   auto existing = snapshot_by_id.find(sri);
+   EOS_ASSERT(existing != snapshot_by_id.end(), chain::snapshot_request_not_found, "Snapshot request not found");
+
+   snapshot_scheduler::snapshot_schedule_result result{{existing->snapshot_request_id}, {existing->block_spacing, existing->start_block_num, existing->end_block_num, existing->snapshot_description}};
```
Comment (greg7mdp) [inline libraries/chain/snapshot_scheduler.cpp:106]:
```suggestion
   snapshot_schedule_result result{{existing->snapshot_request_id}, {existing->block_spacing, existing->start_block_num, existing->end_block_num, existing->snapshot_description}};
```

File: libraries/chain/snapshot_scheduler.cpp [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+      if(sri.block_spacing > 0) {
+         EOS_ASSERT(sri.start_block_num + sri.block_spacing <= sri.end_block_num, chain::invalid_snapshot_request, "Block spacing exceeds defined by start and end range");
+      }
+   }
+
+   _snapshot_requests.emplace(snapshot_scheduler::snapshot_schedule_information{{_snapshot_id++}, {sri.block_spacing, sri.start_block_num, sri.end_block_num, sri.snapshot_description}, {}});
+   x_serialize();
+
+   // returning snapshot_schedule_result
+   return snapshot_scheduler::snapshot_schedule_result{{_snapshot_id - 1}, {sri.block_spacing, sri.start_block_num, sri.end_block_num, sri.snapshot_description}};
```
Comment (greg7mdp) [inline libraries/chain/snapshot_scheduler.cpp:98]:
```suggestion
   return snapshot_schedule_result{{_snapshot_id - 1}, {sri.block_spacing, sri.start_block_num, sri.end_block_num, sri.snapshot_description}};
```

File: libraries/chain/snapshot_scheduler.cpp [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+   snapshot_scheduler::snapshot_schedule_result result{{existing->snapshot_request_id}, {existing->block_spacing, existing->start_block_num, existing->end_block_num, existing->snapshot_description}};
+   _snapshot_requests.erase(existing);
+   x_serialize();
+
+   // returning snapshot_schedule_result
+   return result;
+}
+
+snapshot_scheduler::get_snapshot_requests_result snapshot_scheduler::get_snapshot_requests() {
+   snapshot_scheduler::get_snapshot_requests_result result;
```
Comment (greg7mdp) [inline libraries/chain/snapshot_scheduler.cpp:115]:
```suggestion
   get_snapshot_requests_result result;
```

File: libraries/chain/snapshot_scheduler.cpp [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+      } else {
+         // success, snapshot finalized
+         auto snapshot_info = std::get<snapshot_scheduler::snapshot_information>(result);
+         auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+         auto snapshot_req = snapshot_by_id.find(srid);
+
+         if(snapshot_req != snapshot_by_id.end()) {
+            _snapshot_requests.modify(snapshot_req, [&](auto& p) {
+               auto& pending = p.pending_snapshots;
+               auto it = std::remove_if(pending.begin(), pending.end(), [&snapshot_info](const snapshot_scheduler::snapshot_information& s) { return s.head_block_num <= snapshot_info.head_block_num; });
```
Comment (greg7mdp) [inline libraries/chain/snapshot_scheduler.cpp:172]:
```suggestion
               auto it = std::remove_if(pending.begin(), pending.end(), [&snapshot_info](const snapshot_information& s) { return s.head_block_num <= snapshot_info.head_block_num; });
```

and  more to remove!

File: libraries/chain/snapshot_scheduler.cpp [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+                     "Snapshot creation error: ${details}",
+                     ("details", e.to_detail_string()));
+         } catch(const std::exception& e) {
+            EOS_THROW(snapshot_execution_exception,
+                     "Snapshot creation error: ${details}",
+                     ("details", e.what()));
+         }
+      } else {
+         // success, snapshot finalized
+         auto snapshot_info = std::get<snapshot_scheduler::snapshot_information>(result);
```
Comment (greg7mdp) [inline libraries/chain/snapshot_scheduler.cpp:165]:
```suggestion
         auto snapshot_info = std::get<snapshot_information>(result);
```

File: libraries/chain/snapshot_scheduler.cpp [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+   if(existing != pending_by_id.end()) {
+      // if a snapshot at this block is already pending, attach this requests handler to it
+      pending_by_id.modify(existing, [&next](auto& entry) {
+         entry.next = [prev = entry.next, next](const next_function_variant<snapshot_scheduler::snapshot_information>& res) {
+            prev(res);
+            next(res);
+         };
+      });
+   } else {
+      const auto& pending_path = pending_snapshot<snapshot_scheduler::snapshot_information>::get_pending_path(head_id, _snapshots_dir);
```
Comment (greg7mdp) [inline libraries/chain/snapshot_scheduler.cpp:236]:
```suggestion
      const auto& pending_path = pending_snapshot<snapshot_information>::get_pending_path(head_id, _snapshots_dir);
```

File: plugins/producer_plugin/test/test_snapshot_scheduler.cpp [hunk 8]
```diff
@@ -19,8 +19,8 @@ BOOST_AUTO_TEST_CASE(snapshot_scheduler_test) {
 
    {
       // add/remove test
-      producer_plugin::snapshot_request_information sri1 = {.block_spacing = 100, .start_block_num = 5000, .end_block_num = 10000, .snapshot_description = "Example of recurring snapshot"};
-      producer_plugin::snapshot_request_information sri2 = {.block_spacing = 0, .start_block_num = 5200, .end_block_num = 5200, .snapshot_description = "Example of one-time snapshot"};
+      snapshot_scheduler::snapshot_request_information sri1 = {.block_spacing = 100, .start_block_num = 5000, .end_block_num = 10000, .snapshot_description = "Example of recurring snapshot"};
```
Comment (greg7mdp) [inline plugins/producer_plugin/test/test_snapshot_scheduler.cpp:22]:
```suggestion
      using snapshot_request_information = snapshot_scheduler::snapshot_request_information;
      snapshot_request_information sri1 = {.block_spacing = 100, .start_block_num = 5000, .end_block_num = 10000, .snapshot_description = "Example of recurring snapshot"};
```

and remove following `snapshot_scheduler::`

**Comment by greg7mdp** [comment]
Just an opinion, feel free to ignore. 
If it was me, instead of the three snapshot files in `include/eosio/chain`, I think I would have created a single file `snapshot.hpp` combining these classes. Also I would remove `snapshot_` in front of all the names, but put everything in the namespace `eosio::chain::snapshot`. Also I would move the structs outside of `snapshot_scheduler` and move them at the beginning of the namespace, so they can be used unqualified by all the snapshot code, and the operator>>() and <<() don't have to be templates.

File: libraries/chain/include/eosio/chain/pending_snapshot.hpp [hunk 9]
```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <eosio/chain/controller.hpp>
+#include <eosio/chain/types.hpp>
+
+#include <string>
+
+namespace eosio {
+namespace chain {
```
[the remark this answers is not in the document]
Comment (766C6164) [reply libraries/chain/include/eosio/chain/pending_snapshot.hpp:9]: Thank you, done, looks better

File: libraries/chain/include/eosio/chain/pending_snapshot.hpp [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+namespace bfs = std::filesystem;
+
+template<typename T>
+class pending_snapshot {
+public:
+   using next_t = eosio::chain::next_function<T>;
+
+   pending_snapshot(const chain::block_id_type& block_id, next_t& next, std::string pending_path, std::string final_path)
+       : block_id(block_id), next(next), pending_path(pending_path), final_path(final_path) {}
```
[the remark this answers is not in the document]
Comment (766C6164) [reply libraries/chain/include/eosio/chain/pending_snapshot.hpp:19]: done, thank you

File: libraries/chain/include/eosio/chain/snapshot_db_json.hpp [hunk 11]
```diff
@@ -87,4 +88,5 @@ class snapshot_db_json {
    std::filesystem::path db_path;
```
[the remark this answers is not in the document]
Comment (766C6164) [reply libraries/chain/include/eosio/chain/snapshot_db_json.hpp:88]: done, thank you

**Comment by 766C6164** [comment]
> Just an opinion, feel free to ignore. If it was me, instead of the three snapshot files in `include/eosio/chain`, I think I would have created a single file `snapshot.hpp` combining these classes. Also I would remove `snapshot_` in front of all the names, but put everything in the namespace `eosio::chain::snapshot`. Also I would move the structs outside of `snapshot_scheduler` and move them at the beginning of the namespace, so they can be used unqualified by all the snapshot code, and the operator>>() and <<() don't have to be templates.

Thank you, it all make sense. I partially incorporated - eliminated separate db header and merged it in with scheduler, since it always used by it anyways. That allowed to drop templates for operators as well.

File: plugins/producer_plugin/test/test_snapshot_scheduler.cpp [hunk 12]
```diff
@@ -105,9 +108,9 @@ BOOST_AUTO_TEST_CASE(snapshot_scheduler_test) {
...
-         producer_plugin::snapshot_request_information sri2 = {.block_spacing = 5000, .start_block_num = 100000, .end_block_num = 300000, .snapshot_description = "Example of recurring snapshot 2 that will never happen"};
-         producer_plugin::snapshot_request_information sri3 = {.block_spacing = 2, .start_block_num = 0, .end_block_num = 3, .snapshot_description = "Example of recurring snapshot 3 that will expire"};
+         snapshot_scheduler::snapshot_request_information sri1 = {.block_spacing = 8, .start_block_num = 1, .end_block_num = 300000, .snapshot_description = "Example of recurring snapshot 1"};
+         snapshot_scheduler::snapshot_request_information sri2 = {.block_spacing = 5000, .start_block_num = 100000, .end_block_num = 300000, .snapshot_description = "Example of recurring snapshot 2 that will never happen"};
+         snapshot_scheduler::snapshot_request_information sri3 = {.block_spacing = 2, .start_block_num = 0, .end_block_num = 3, .snapshot_description = "Example of recurring snapshot 3 that will expire"};
```
Comment (greg7mdp) [inline plugins/producer_plugin/test/test_snapshot_scheduler.cpp:113]:
```suggestion
         snapshot_request_information sri1 = {.block_spacing = 8, .start_block_num = 1, .end_block_num = 300000, .snapshot_description = "Example of recurring snapshot 1"};
         snapshot_request_information sri2 = {.block_spacing = 5000, .start_block_num = 100000, .end_block_num = 300000, .snapshot_description = "Example of recurring snapshot 2 that will never happen"};
         snapshot_request_information sri3 = {.block_spacing = 2, .start_block_num = 0, .end_block_num = 3, .snapshot_description = "Example of recurring snapshot 3 that will expire"};
```

**Review by greg7mdp — APPROVED** [review]

**Comment by greg7mdp** [comment]
> > Just an opinion, feel free to ignore. If it was me, instead of the three snapshot files in `include/eosio/chain`, I think I would have created a single file `snapshot.hpp` combining these classes. Also I would remove `snapshot_` in front of all the names, but put everything in the namespace `eosio::chain::snapshot`. Also I would move the structs outside of `snapshot_scheduler` and move them at the beginning of the namespace, so they can be used unqualified by all the snapshot code, and the operator>>() and <<() don't have to be templates.
> 
> Thank you, it all make sense. I partially incorporated - eliminated separate db header and merged it in with scheduler, since it always used by it anyways. That allowed to drop templates for operators as well.

Thanks Vlad, looks good!

File: plugins/producer_plugin/test/test_snapshot_scheduler.cpp (hunk 12 above)
Comment (766C6164) [reply plugins/producer_plugin/test/test_snapshot_scheduler.cpp]: Thank you, missed that, fixed now!

**Review by linh2931 — COMMENTED** [review]
Ho did you move the files? The moved files were shown deleted and recreated fresh new. The commit history was lost. I thought Github would show a file moved from another location and keep the history. I usually use command line "mv" and github knows I am moving and keeps the commits.

**Review by linh2931 — APPROVED** [review]
Vlad did some research and did not figure out better ways to save the history. As most files are not worth having history saved -  some were hpp/cpp and became just header, another file (snapshot_db_json) gone completely and its encapsulated into another one, we agreed it was not worth the effort.

**Pull request merged**